algorithms/src/polycommit/data_structures/powers.rs

```diff
@@ -0,0 +1,375 @@
+        m.insert(1 << 27, String::from("https://f002.backblazeb2.com/file/aleo-snarkos/srs/powers_of_g_27"));
+        m.insert(1 << 28, String::from("https://f002.backblazeb2.com/file/aleo-snarkos/srs/powers_of_g_28"));
+        m
+    };
+    static ref BASE_POWERS: &'static [u8] = include_bytes!("./powers_of_g_15");
+    static ref POWERS_TIMES_GAMMA_G: &'static [u8] = include_bytes!("./gamma_powers");
+}
+
+// Amount of powers contained in `POWERS_TIMES_GAMMA_G`.
+const NUM_POWERS_TIMES_GAMMA_G: usize = 84;
+
+/// An abstraction over a vector of powers of G, meant to reduce
+/// memory burden when handling universal setup parameters.
+#[derive(Derivative)]
+#[derivative(Debug(bound = ""))]
+pub struct PowersOfG<E: PairingEngine> {
+    /// Filepath of the powers we're using.
+    file_path: String,
+    /// A handle to the file on disk containing the powers of G.
+    /// The handle is guarded to avoid read/write conflicts with potential
+    /// clones.
+    #[derivative(Debug = "ignore")]
+    file: Arc<RwLock<File>>,
+    /// Group elements of the form `{ \beta^i \gamma G }`, where `i` ranges from 0 to `degree`.
+    /// These are used for hiding.
+    pub powers_of_beta_times_gamma_g: Arc<RwLock<BTreeMap<usize, E::G1Affine>>>,
+    /// The degree up to which we have powers.
+    /// Mostly just used for ensuring we download in the proper order,
+    /// length checks are actually done by checking file metadata.
+    degree: Arc<RwLock<usize>>,
+    _phantom_data: PhantomData<E>,
+}
+
+impl<E: PairingEngine> Default for PowersOfG<E> {
+    fn default() -> Self {
+        // This basically should never fail, hence the expect call.
+        Self::new(DEFAULT_PATH.clone()).expect("could not create default powers of G")
+    }
+}
+
+impl<E: PairingEngine> Clone for PowersOfG<E> {
+    fn clone(&self) -> Self {
+        Self::new(PathBuf::from(self.file_path.clone())).expect("could not clone powers of g")
+    }
+}
+
+impl<E: PairingEngine> ToBytes for PowersOfG<E> {
+    fn write_le<W: Write>(&self, mut writer: W) -> std::io::Result<()> {
+        // Serialize `powers_of_beta_times_gamma_g`.
+        (self.powers_of_beta_times_gamma_g.read().len() as u32).write_le(&mut writer)?;
+        for (key, power_of_gamma_g) in &*self.powers_of_beta_times_gamma_g.read() {
+            (*key as u32).write_le(&mut writer)?;
+            power_of_gamma_g.write_le(&mut writer)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl<E: PairingEngine> FromBytes for PowersOfG<E> {
+    fn read_le<R: Read>(mut reader: R) -> std::io::Result<Self> {
+        // Deserialize `powers_of_beta_times_gamma_g`.
+        let mut powers_of_beta_times_gamma_g = BTreeMap::new();
+        let powers_of_gamma_g_num_elements: u32 = FromBytes::read_le(&mut reader)?;
+        for _ in 0..powers_of_gamma_g_num_elements {
+            let key: u32 = FromBytes::read_le(&mut reader)?;
+            let power_of_gamma_g: E::G1Affine = FromBytes::read_le(&mut reader)?;
```

Comment:
Doesn't this use the default impl which performs subgroup and on-curve checks?

Reply:
Hm let me revive #459; it enables unchecked deserialization which enables omitting on-curve and in-subgroup checks when deserializing from trusted sources.